File: src/033SearchInRotatedSortedArray/033SearchInRotatedSortedArray.go
package main

import "fmt"

func search(nums []int, target int) int {
	n := len(nums)
	lo := 0
	hi := n - 1
	for lo <= hi {
		mi := lo + (hi-lo)/2
		if nums[mi] == target {
			return mi
		} else if nums[mi] >= nums[hi] {
			if nums[mi] > target && nums[lo] <= target {
				hi = mi - 1
			} else {
				lo = mi + 1
			}
		} else {
			if nums[mi] < target && nums[hi] >= target {
				lo = mi + 1
			} else {
				hi = mi - 1
			}
		}
	}

	return -1
}

func main() {
	fmt.Printf("%v\n", search([]int{4, 5, 6, 7, 0, 1, 2}, 3))
	fmt.Printf("%v\n", search([]int{4, 5, 6, 7, 0, 1, 2}, 4))
	fmt.Printf("%v\n", search([]int{4, 5, 6, 7, 0, 1, 2}, 7))
}

File: src/033SearchInRotatedSortedArray/033SearchInRotatedSortedArray.rs
impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let (mut lo, mut hi) = (0 as i32, nums.len() as i32 - 1);
        while lo <= hi {
            let mi = lo + (hi-lo)/2;
            if nums[mi as usize] == target {
                return mi;
            } else if nums[mi as usize] >= nums[lo as usize] {
                if target < nums[mi as usize] && target >= nums[lo as usize] {
                    hi = mi - 1;
                } else {
                    lo = mi + 1;
                }
            } else {
                if target > nums[mi as usize] && target <= nums[hi as usize] {
                    lo = mi + 1;
                } else {
                    hi = mi - 1;
                }
            }
        }
        
        return -1;
    }
}
